pydrive2/fs/spec.py

```diff
@@ -473,6 +473,11 @@ def find(self, path, detail=False, **kwargs):
         bucket, base = self.split_path(path)
 
         seen_paths = set()
+        cached = base in self._ids_cache["dirs"]
+        if not cached:
+            dir_ids = self._path_to_item_ids(base)
+            self._cache_path_id(base, *dir_ids)
```

Comment:
my concern here is that it's not thread safe most likely (and we run multiple find operation in parallel in DVC AFAIK in certain cases). Before, we had only two important (for DVC) place - initial cache creation and create dir - both are wrapper with locks.

With this change, when we are now relying on `find` updating the cache here and below, things become more complicated for us.

Reply:
Could you please elaborate what your concerns are?

https://github.com/iterative/PyDrive2/blob/42c816028193fdf3c86bf88aecb333076ba88202/pydrive2/fs/spec.py#L269-L272

These operations are append-only/overwrite, which should be threadsafe.

Reply:
yep, may be that's fine. Most likely that lock on top of the cache properly was done to instantiate it once, and for the directory to make sure that we don't create it twice.

if you don't see any other obvious thread safety concerns (e.g. is it a problem that the list can become a 200 items long (I wonder why don't we use a set here?)

Reply:
> if you don't see any other obvious thread safety concerns (e.g. is it a problem that the list can become a 200 items long (I wonder why don't we use a set here?)

I'm not sure why it is a list, but we should prevent it from adding if it's already cached.
Right now, it can have duplicated ids if you call `find()` too many times.

Btw, what's the use of a list? Is it possible for a path to have multiple ids?

Reply:
Okay, it looks like it can have multiple ids, but I am not sure what they are supposed to be.

Reply:
I found some information on https://github.com/iterative/dvc/pull/2551#discussion_r345694266.

It seems the cache was a list from early on. Will look into if we can make it a set in a separate PR.